extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,1263 @@
+  }
+
+  @Override
+  public void stop(boolean stopGracefully)
+  {
+    Preconditions.checkState(started, "not started");
+
+    log.info("Beginning shutdown of KafkaSupervisor[%s]", dataSource);
+
+    try {
+      timer.cancel(); // stop recurring executions
+
+      Optional<TaskRunner> taskRunner = taskMaster.getTaskRunner();
+      if (taskRunner.isPresent()) {
+        taskRunner.get().unregisterListener(supervisorId);
+      }
+
+      // Stopping gracefully will synchronize the end offsets of the tasks and signal them to publish, and will block
+      // until the tasks have acknowledged or timed out. We want this behavior when we're explicitly shut down through
+      // the API, but if we shut down for other reasons (e.g. we lose leadership) we want to just stop and leave the
+      // tasks as they are.
+      if (stopGracefully) {
+        log.info("Stopping gracefully, signalling managed tasks to complete and publish");
+        synchronized (stopLock) {
+          notices.add(new ShutdownNotice());
+          while (!stopped) {
+            stopLock.wait();
+          }
+        }
+        log.info("Shutdown notice handled");
+      }
+
+      exec.shutdownNow();
+      consumer.close();
+      started = false;
+
+      log.info("KafkaSupervisor[%s] has stopped", dataSource);
+    }
+    catch (Exception e) {
+      log.makeAlert(e, "Exception stopping KafkaSupervisor[%s]", dataSource)
+         .emit();
+    }
+  }
+
+  @Override
+  public SupervisorReport getStatus()
+  {
+    return generateReport(true);
+  }
+
+  public void possiblyRegisterListener()
+  {
+    // getTaskRunner() sometimes fails if the task queue is still being initialized so retry later until we succeed
+
+    if (listenerRegistered) {
+      return;
+    }
+
+    Optional<TaskRunner> taskRunner = taskMaster.getTaskRunner();
+    if (taskRunner.isPresent()) {
+      taskRunner.get().registerListener(
+          new TaskRunnerListener()
+          {
+            @Override
+            public String getListenerId()
+            {
+              return supervisorId;
+            }
+
+            @Override
+            public void locationChanged(final String taskId, final TaskLocation newLocation)
+            {
+              // do nothing
+            }
+
+            @Override
+            public void statusChanged(String taskId, TaskStatus status)
+            {
+              notices.add(new RunNotice());
+            }
+          }, MoreExecutors.sameThreadExecutor()
+      );
+
+      listenerRegistered = true;
+    }
+  }
+
+  private interface Notice
+  {
+    void handle();
+  }
+
+  private class RunNotice implements Notice
+  {
+    @Override
+    public void handle()
+    {
+      long nowTime = System.currentTimeMillis();
+      if (nowTime - lastRunTime < MAX_RUN_FREQUENCY_MILLIS) {
+        return;
+      }
+      lastRunTime = nowTime;
+
+      runInternal();
```

Comment:
@gianm ah, I actually had it info log a summary of the tasks it was managing but then upon looking at the logs thought it was getting too spammy so I removed it yesterday. The log messages looked like this:
```
2016-05-02T22:01:34,527 INFO [KafkaSupervisor-github-v2] io.druid.indexing.kafka.supervisor.KafkaSupervisor - {id='github-v2', generationTime=2016-05-02T22:01:34.527Z, payload={dataSource='github-v2', topic='github', partitions=3, replicas=1, durationSeconds=3600, active=[{id='index_kafka_github-v2_93416522dbb8099_olhpaebd', startTime=2016-05-02T21:33:28.394Z, remainingSeconds=1913}, {id='index_kafka_github-v2_965719e6f2b23f1_pkeljcng', startTime=2016-05-02T21:32:45.640Z, remainingSeconds=1871}], publishing=[]}}
2016-05-02T22:01:34,548 INFO [KafkaSupervisor-twitter-v2] io.druid.indexing.kafka.supervisor.KafkaSupervisor - {id='twitter-v2', generationTime=2016-05-02T22:01:34.548Z, payload={dataSource='twitter-v2', topic='twitter', partitions=3, replicas=2, durationSeconds=3600, active=[{id='index_kafka_twitter-v2_1c4fab566153c68_kbpjknfe', startTime=2016-05-02T21:32:28.397Z, remainingSeconds=1853}, {id='index_kafka_twitter-v2_1c4fab566153c68_diiajjaa', startTime=2016-05-02T21:32:31.193Z, remainingSeconds=1856}, {id='index_kafka_twitter-v2_53f8a23dfcdfa3a_mhofcgeb', startTime=2016-05-02T21:26:24.668Z, remainingSeconds=1490}, {id='index_kafka_twitter-v2_53f8a23dfcdfa3a_dblegedd', startTime=2016-05-02T21:26:22.060Z, remainingSeconds=1487}], publishing=[]}}
2016-05-02T22:01:54,527 INFO [KafkaSupervisor-github-v2] io.druid.indexing.kafka.supervisor.KafkaSupervisor - {id='github-v2', generationTime=2016-05-02T22:01:54.527Z, payload={dataSource='github-v2', topic='github', partitions=3, replicas=1, durationSeconds=3600, active=[{id='index_kafka_github-v2_93416522dbb8099_olhpaebd', startTime=2016-05-02T21:33:28.394Z, remainingSeconds=1893}, {id='index_kafka_github-v2_965719e6f2b23f1_pkeljcng', startTime=2016-05-02T21:32:45.640Z, remainingSeconds=1851}], publishing=[]}}
```
What do you think?